src/python/pants/reporting/json_reporter.py

```diff
@@ -0,0 +1,145 @@
+from pants.base.build_environment import get_buildroot
+from pants.reporting.reporter import Reporter
+from pants.util.dirutil import safe_mkdir
+
+
+class JsonReporter(Reporter):
+  """Json reporting to files."""
+
+  _log_level_str = [
+    'FATAL',
+    'ERROR',
+    'WARN',
+    'INFO',
+    'DEBUG',
+  ]
+
+  # JSON reporting settings.
+  #   json_dir: Where the report files go.
+  Settings = namedtuple('Settings', Reporter.Settings._fields + ('json_dir',))
+
+  def __init__(self, run_tracker, settings):
+    super(JsonReporter, self).__init__(run_tracker, settings)
+    # The main report, and associated tool outputs, go under this dir.
+    self._json_dir = settings.json_dir
+
+    self._buildroot = get_buildroot()
+    self._report_file = None
+    self._output_files = defaultdict(dict)
+
+    # Results of the state of the build
+    self._results = {}
+    # Workunit stacks partitioned by thread
+    self._stack_per_thread = defaultdict(list)
+
+  def report_path(self):
+    """The path to the main report file."""
+
+    return os.path.join(self._json_dir, 'build.json')
+
+  def open(self):
+    """Implementation of Reporter callback."""
+
+    safe_mkdir(os.path.dirname(self._json_dir))
+    self._report_file = open(self.report_path(), 'w')
+
+  def close(self):
+    """Implementation of Reporter callback."""
+
+    if os.path.exists(self._json_dir):
+      self._report_file.write(json.dumps(
+        {
+          'workunits': self._results,
+          'artifact_cache_stats': self.run_tracker.artifact_cache_stats.get_all(),
+        }
+      ))
+      self._report_file.flush()
+    self._report_file.close()
+    for files in self._output_files.values():
+      for f in files.values():
+        f.close()
+
+  def start_workunit(self, workunit):
+    """Implementation of Reporter callback."""
+
+    workunit_data = {
+      'name': workunit.name,
+      'id': str(workunit.id),
+      'parent_name': workunit.parent.name if workunit.parent else '',
+      'parent_id': str(workunit.parent.id) if workunit.parent else '',
+      'labels': list(workunit.labels),
+      'cmd': workunit.cmd or '',
+      'start_time': workunit.start_time,
+      'outputs': defaultdict(str),
+      'children': [],
+      'log_entries': [],
+    }
+
+    root_id = str(workunit.root().id)
+
+    if not root_id in self._stack_per_thread:
+      self._stack_per_thread[root_id].append(workunit_data)
+      self._results[root_id] = workunit_data
+    else:
+      self._stack_per_thread[root_id][-1]['children'].append(workunit_data)
+      self._stack_per_thread[root_id].append(workunit_data)
+
+  def end_workunit(self, workunit):
+    """Implementation of Reporter callback."""
+
+    additional_data = {
+      'outcome': workunit.outcome_string(workunit.outcome()),
+      'end_time': workunit.end_time,
+      'unaccounted_time': workunit.unaccounted_time(),
+    }
+
+    root_id = str(workunit.root().id)
+
+    self._stack_per_thread[root_id][-1].update(additional_data)
```

Comment:
It might be good to note that this is modifying something that is also pointed to from another data structure. I found the `update`, `pop` here confusing before I figured that out.